overview document

rtertiaer opened the pull request:
This PR adds an overview document that sketches out the flow of data using this support tunnel setup. I'd appreciate input on a couple of things:

* do I have any blind spots that I'm not explaining in enough detail?
* Is it reasonably well written?
* Is there anything amiss with this design? lol

linknum23 commented:
I'll check this out in depth tomorrow. My instincts tell me that you are missing one or two diagrams.

rtertiaer commented:
good call, I'll add a diagram. (@linknum23 I'd love your further input, but don't feel like you need to provide it - I've tapped others.)

klay2000 commented on docs/overview.md:
```diff
@@ -0,0 +1,34 @@
+
+## Tunnel instantiation
+
+The device checks back in and finds that the tunnel has been approved and there is a cloud server waiting for it. It starts its own WireGuard tunnel using the public IP and public key of the cloud server (fetched from the API). It also generates an ephemeral support user with a random prefix, unwraps the SSH `authorized_keys` entry from the NaCL box and creates it, and gives this user sudo permis ...
```
how frequently is pretty frequently?

klay2000 reviewed (commented):
explain where numbers for some of these times are kept, or put numbers directly here
It all makes sense to me, or at least enough sense that I _think_ I understand it

linknum23 commented on docs/overview.md:
```diff
@@ -0,0 +1,34 @@
+# Overview
+
+This document presents a high level overview of the flow of information and steps taken to establish a support tunnel, in technical detail. It is intended to show at a high level that this implementation is securely designed.
```
I would suggest describing the overall intent of the support tunnel: Something that only a user can initiate that doesn't expose them to any additional security risks.

linknum23 commented on docs/overview.md:
```diff
@@ -0,0 +1,34 @@
+
+When a request is initiated from a `device`, some initial values are generated - an Ed25519 public+private keypair and a small private subnet that isn't already present in the device's routing table. It saves these values into a local SQLite DB.
+
+Then a `device` requests a tunnel from the `api`, running in the cloud. This request contains the public key and the private network subnet to store in the `api`'s database, and will return a JWT that contains a `tunnel_id` and an expiry time. From this point on, the `device` can only interact with the API by returning this JWT as a bearer token, identifying it as this particular tunnel.
```
JWT could use an initial full acronym. I had to look it up.

linknum23 commented on docs/overview.md:
```diff
@@ -0,0 +1,34 @@
+
+When a support request is received with these details, the tunnel is then approved or denied. When a tunnel is approved, the support engineer provides the `tunnel_id` and the `preshared_key` provided by the user to a local CLI application. This local application will start a server in the cloud and configure it using [Fabric](https://www.fabfile.org/).
+
+The cloud server will generate its own Ed25519 public+private keypair, and store the public IP and public key in the API's database. The cloud server is configured by the local CLI app to start a WireGuard tunnel, using the local public+private keypair, the preshared key, and the `device`'s public key. A `support` user is created and an SSH key is generated to permit keyed access to the remote de ...
```
What is NaCl ?

rtertiaer commented:
I've implemented a diagram, added some links and explainer text where I had blind spots, and linked to various parts where constants are defined. I think I've implemented all feedback here

linknum23 commented:
Just a heads up that diagram doesn't look great in dark mode.

linknum23 reviewed (approved):
This looks good to me. I only request that you change the theme of that diagram the arrows look ridiculous. The tunnel server box should probably be moved up, keeping the dashed lines.

rtertiaer commented:
and here I thought it looked cute :sob: 

revisiting it, I was able to clean up the diagram a bunch too; it's a lot more organized and straightforward.

linknum23 commented:
That looks a lot better.
